File: simsys_tokens/store.py
"""The token store: schema, migrations, and row access.

ensure_schema() is the ONE creation/migration routine. install_tokens() and every
CLI subcommand call it — on a fresh install the bootstrap token is minted before
the service has ever started, so the CLI must be able to create the schema too,
and two independent creation paths would drift.
"""

from __future__ import annotations

import contextlib
import os
import sqlite3
import threading
from datetime import datetime, timezone
from pathlib import Path

from .errors import TokenError
from .hashing import (
    handle_of,
    mint_token,
    token_hash,
    validate_handle,
    validate_role,
    validate_service,
)

SCHEMA_VERSION = 2

_DDL = """
CREATE TABLE IF NOT EXISTS tokens (
  token_sha256 TEXT PRIMARY KEY,
  service      TEXT NOT NULL,
  label        TEXT NOT NULL,
  role         TEXT NOT NULL,
  priority     INTEGER,
  rate_limit   TEXT,
  expires_at   TEXT,
  created_at   TEXT NOT NULL,
  created_by   TEXT NOT NULL,
  last_used_at TEXT,
  revoked_at   TEXT,
  revoked_by   TEXT
);
CREATE INDEX IF NOT EXISTS idx_tokens_live ON tokens(service, revoked_at);
-- Expression index for resolve_handle(): without it the substr() predicate
-- cannot use the PRIMARY KEY and degrades to a full scan. SQLite >= 3.9
-- supports this; Python 3.10 bundles >= 3.37. IF NOT EXISTS so existing
-- stores gain it on the next ensure_schema() run.
CREATE INDEX IF NOT EXISTS idx_tokens_handle ON tokens(service, substr(token_sha256, 1, 16));
-- Live-label uniqueness enforced by the DATABASE, not by check-then-insert.
-- Two concurrent creates can both pass a live_label_exists() check and then both
-- insert; a partial unique index makes that impossible rather than unlikely.
CREATE UNIQUE INDEX IF NOT EXISTS idx_tokens_live_label
  ON tokens(service, label) WHERE revoked_at IS NULL;
-- Single-row by construction: a bare (version INTEGER) table lets a double-run
-- leave two rows and makes SELECT version non-deterministic.
CREATE TABLE IF NOT EXISTS _schema_version (
  id      INTEGER PRIMARY KEY CHECK (id = 1),
  version INTEGER NOT NULL
);
"""

# Columns added to `tokens` after v1, with the migration that introduces each.
# `CREATE TABLE IF NOT EXISTS` never alters an existing table, so a bare edit to
# _DDL silently no-ops on adopted stores — every such column needs a step here.
_COLUMN_MIGRATIONS = {
    "expires_at": "ALTER TABLE tokens ADD COLUMN expires_at TEXT",
}

MAX_LIMIT = 500
# Bound for the per-process throttle maps. Thousands of tokens fleet-wide means
# a few thousand entries max in practice; 5000 with oldest-first eviction keeps
# memory flat if a scanner hammers random digests.
TOUCH_MAP_MAX = 5000


def _utcnow() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def normalize_expires_at(value) -> str | None:
    """Return a UTC ISO-8601 timestamp, or None. Raises ValueError on garbage.

    Normalising to one format is load-bearing: expiry is compared as a string,
    so ``2026-01-01T00:00:00+00:00`` and a naive ``2026-01-01T00:00:00`` would
    otherwise compare against each other as different-length text.
    """
    if value is None:
        return None
    if not isinstance(value, str) or not value.strip():
        raise ValueError("expires_at must be an ISO-8601 timestamp or null")
    try:
        parsed = datetime.fromisoformat(value)
    except ValueError:
        raise ValueError(f"expires_at is not an ISO-8601 timestamp: {value!r}") from None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc).isoformat(timespec="seconds")


def is_expired(expires_at: str | None, now: str | None = None) -> bool:
    """True iff *expires_at* is set and not in the future. Strings compare
    lexicographically because every stored timestamp is normalised to UTC."""
    if not expires_at:
        return False
    return expires_at <= (now or _utcnow())


@contextlib.contextmanager
def connect(db_path: str | Path):
    """Transaction AND lifetime scoped.

    sqlite3's own connection context manager commits or rolls back but NEVER
    closes, so a bare `with sqlite3.connect(...)` leaks a handle per call — which
    becomes file-descriptor growth and "database is locked" under the
    multi-worker adopters this package is meant to serve. WAL and a busy timeout
    are set for the same reason: several workers share one file.
    """
    conn = sqlite3.connect(str(db_path), timeout=5.0)
    try:
        conn.row_factory = sqlite3.Row
        conn.execute("PRAGMA journal_mode = WAL")
        conn.execute("PRAGMA busy_timeout = 5000")
        conn.execute("PRAGMA foreign_keys = ON")
        with conn:  # commit on success, rollback on exception
            yield conn
    finally:
        conn.close()


def _restrict_permissions(path: Path) -> None:
    """Best-effort chmod 0600 on the store and its WAL sidecars.

    The store holds digests, labels and actors — not raw secrets — so this is
    hygiene rather than a security boundary. Best-effort because Windows has no
    POSIX mode and some filesystems reject chmod; a failure here must not stop
    an app from starting.
    """
    for candidate in (path, Path(f"{path}-wal"), Path(f"{path}-shm")):
        try:
            if candidate.exists():
                os.chmod(candidate, 0o600)
        except OSError:
            pass


def ensure_schema(db_path: str | Path, *, create: bool) -> None:
    path = Path(db_path)
    if not path.exists():
        if not create:
            raise FileNotFoundError(
                f"{path} does not exist. Pass --init to create it, and check the path "
                f"matches the one the app passes to install_tokens()."
            )
        path.parent.mkdir(parents=True, exist_ok=True)
    with connect(path) as conn:
        # Check the version BEFORE applying any DDL. Running this package's
        # schema against a store written by a newer package and *then* refusing
        # to downgrade is the wrong order: the damage is already done by the
        # time the error is raised.
        existing = (
            conn.execute("SELECT version FROM _schema_version WHERE id = 1").fetchone()
            if _table_exists(conn, "_schema_version")
            else None
        )
        if existing is not None and existing["version"] > SCHEMA_VERSION:
            raise TokenError(
                f"store at {path} is schema v{existing['version']}, newer than this "
                f"package (v{SCHEMA_VERSION}). Refusing to downgrade — no DDL applied."
            )
        if existing is not None and existing["version"] < SCHEMA_VERSION:
            _migrate(conn, existing["version"])
        # executescript() implicitly COMMITs before it runs, so it is NOT inside
        # the surrounding transaction — deliberate here, since DDL is idempotent
        # and each migration step above is separately committed.
        conn.executescript(_DDL)
        # Upsert, not INSERT OR IGNORE: an upgraded store already has the id=1
        # row, and a plain ignore would leave its version at the old value
        # forever, re-running the migration on every start.
        conn.execute(
            "INSERT INTO _schema_version (id, version) VALUES (1, ?)"
            " ON CONFLICT(id) DO UPDATE SET version = excluded.version",
            (SCHEMA_VERSION,),
        )
    _restrict_permissions(path)


def _migrate(conn, from_version: int) -> None:
    """Incremental migration: bring *from_version* up to SCHEMA_VERSION.

    Every step is guarded by a column/table check so a store that was created at
    the current version — or one where a previous run half-applied — converges
    rather than failing on a duplicate column.
    """
    if from_version < 2:
        for column, ddl in _COLUMN_MIGRATIONS.items():
            if not _column_exists(conn, "tokens", column):
                conn.execute(ddl)
    if from_version < 1:  # unreachable; guards a hand-edited version row
        raise TokenError(f"store claims schema v{from_version}; no path from there")


def _table_exists(conn, name: str) -> bool:
    return (
        conn.execute(
            "SELECT 1 FROM sqlite_master WHERE type='table' AND name=?", (name,)
        ).fetchone()
        is not None
    )


def _column_exists(conn, table: str, column: str) -> bool:
    return any(row["name"] == column for row in conn.execute(f"PRAGMA table_info({table})"))


class NoSuchHandle(TokenError):
    """Raised with a sentence, not a bare handle — the CLI prints str(exc)
    directly, and `error: 0000000000000000` tells the operator nothing."""


class AmbiguousHandle(TokenError):
    pass


class Store:
    """Service-scoped row access. Every query carries `service`."""

    def __init__(
        self,
        db_path,
        service: str,
        *,
        max_limit: int = MAX_LIMIT,
        touch_map_max: int = TOUCH_MAP_MAX,
    ):
        validate_service(service)
        self.db_path = db_path
        self.service = service
        self.max_limit = max_limit
        self.touch_map_max = touch_map_max
        # Per-store, not module-global: a global throttle map keyed only on the
        # digest is shared across every service in the process and never shrinks.
        # Guarded by _touch_lock: adapters run threaded (Flask) and async workers
        # share the Store, so an unlocked dict check-then-set races.
        self.last_touch: dict[str, float] = {}
        self.last_auth_failed: dict[str, float] = {}
        self._touch_lock = threading.Lock()

    def _touch_note(self, mapping: dict[str, float], key: str, now: float) -> None:
        """Record `key` at `now`, evicting oldest-first when over the bound.

        pop-then-set, not a bare assignment: re-setting an existing key would
        otherwise keep its original insertion position, so a key that just
        re-emitted could be evicted as if it were the oldest.
        """
        mapping.pop(key, None)
        mapping[key] = now
        while len(mapping) > self.touch_map_max:
            mapping.pop(next(iter(mapping)), None)

    def mint(self, role, label, created_by, priority=None, rate_limit=None, expires_at=None):
        validate_role(role)
        raw = mint_token(self.service, role)
        handle = self.insert_digest(
            token_hash(raw), role, label, created_by, priority, rate_limit, expires_at
        )
        return raw, handle

    def insert_digest(
        self, digest, role, label, created_by, priority=None, rate_limit=None, expires_at=None
    ):
        """Plain INSERT — raises sqlite3.IntegrityError on any conflict.

        Deliberately NOT `INSERT OR IGNORE`: that would suppress the live-label
        unique-index violation as well as the primary-key one, silently dropping
        a colliding import instead of reporting it. Import idempotence comes from
        the         caller's find_any() check, not from the insert mode.
        """
        validate_role(role)
        with connect(self.db_path) as conn:
            conn.execute(
                "INSERT INTO tokens (token_sha256, service, label, role, priority,"
                " rate_limit, expires_at, created_at, created_by)"
                " VALUES (?,?,?,?,?,?,?,?,?)",
                (
                    digest,
                    self.service,
                    label,
                    role,
                    priority,
                    rate_limit,
                    normalize_expires_at(expires_at),
                    _utcnow(),
                    created_by,
                ),
            )
        return handle_of(digest)

    def lookup_live(self, digest, now: str | None = None):
        """The auth lookup: service-scoped, not revoked, not expired."""
        with connect(self.db_path) as conn:
            return conn.execute(
                "SELECT * FROM tokens WHERE token_sha256 = ? AND service = ?"
                " AND revoked_at IS NULL"
                " AND (expires_at IS NULL OR expires_at > ?)",
                (digest, self.service, now or _utcnow()),
            ).fetchone()

    def find_any(self, digest):
        with connect(self.db_path) as conn:
            return conn.execute(
                "SELECT * FROM tokens WHERE token_sha256 = ? AND service = ?",
                (digest, self.service),
            ).fetchone()

    def list_rows(self, include_revoked=False, limit=None, label=None):
        # Clamp BOTH ends. A negative limit reaches SQLite as `LIMIT -1`, which
        # means "no limit" and returns every row, while `rows[:limit]` with a
        # negative slice silently drops the tail and miscomputes has_more.
        limit = self.max_limit if limit is None else int(limit)
        limit = max(1, min(limit, self.max_limit))
        clause = "" if include_revoked else " AND revoked_at IS NULL"
        params: list = [self.service]
        if label is not None:
            clause += " AND label = ?"
            params.append(label)
        with connect(self.db_path) as conn:
            rows = conn.execute(
                f"SELECT * FROM tokens WHERE service = ?{clause}"
                " ORDER BY created_at, rowid LIMIT ?",
                (*params, limit + 1),
            ).fetchall()
        return (rows[:limit], len(rows) > limit)

    def resolve_handle(self, handle):
        validate_handle(handle)
        with connect(self.db_path) as conn:
            rows = conn.execute(
                "SELECT * FROM tokens WHERE service = ? AND substr(token_sha256,1,16) = ?",
                (self.service, handle),
            ).fetchall()
        if not rows:
            raise NoSuchHandle(f"no token with handle {handle!r} in service {self.service!r}")
        if len(rows) > 1:
            raise AmbiguousHandle(
                f"handle {handle!r} matches {len(rows)} tokens in service "
                f"{self.service!r}; use a longer prefix from the listing"
            )
        return rows[0]

    def revoke(self, handle, revoked_by):
        """Returns (row, changed). `changed` is False when it was already revoked.

        The guard is in the UPDATE's WHERE clause, not a preceding read: a
        check-then-act pair lets two concurrent revokes both pass the check and
        the second overwrite the first's revoked_by, losing the record of who
        actually killed the token.
        """
        row = self.resolve_handle(handle)
        with connect(self.db_path) as conn:
            cur = conn.execute(
                "UPDATE tokens SET revoked_at = ?, revoked_by = ?"
                " WHERE token_sha256 = ? AND service = ? AND revoked_at IS NULL",
                (_utcnow(), revoked_by, row["token_sha256"], self.service),
            )
            changed = cur.rowcount == 1
        return self.resolve_handle(handle), changed

    def update_row(self, handle, *, label=None, priority=..., rate_limit=..., expires_at=...):
        row = self.resolve_handle(handle)
        sets, params = [], []
        if label is not None:
            sets.append("label = ?")
            params.append(label)
        if priority is not ...:
            sets.append("priority = ?")
            params.append(priority)
        if rate_limit is not ...:
            sets.append("rate_limit = ?")
            params.append(rate_limit)
        if expires_at is not ...:
            sets.append("expires_at = ?")
            params.append(normalize_expires_at(expires_at))
        if sets:
            # `AND service = ?` like every other query in this class. Safe
            # without it only because token_sha256 is the primary key — an
            # invariant a future schema change could break silently.
            params.extend([row["token_sha256"], self.service])
            with connect(self.db_path) as conn:
                conn.execute(
                    f"UPDATE tokens SET {', '.join(sets)} WHERE token_sha256 = ? AND service = ?",
                    params,
                )
        return self.resolve_handle(handle)

    def update_policy(self, digest, priority, rate_limit):
        """Policy-only update, keyed on the digest. Used by the importer's
        re-import path — every row write goes through Store, so there is one
        place to audit for SQL."""
        with connect(self.db_path) as conn:
            conn.execute(
                "UPDATE tokens SET priority = ?, rate_limit = ?"
                " WHERE token_sha256 = ? AND service = ?",
                (priority, rate_limit, digest, self.service),
            )

    def touch_last_used(self, digest):
        # No explicit commit: connect() commits on clean exit, so an extra
        # conn.commit() here is redundant (and masks the context-manager contract).
        with connect(self.db_path) as conn:
            conn.execute(
                "UPDATE tokens SET last_used_at = ? WHERE token_sha256 = ? AND service = ?",
                (_utcnow(), digest, self.service),
            )

    def live_label_exists(self, label, *, excluding_digest=None):
        sql = "SELECT 1 FROM tokens WHERE service = ? AND label = ? AND revoked_at IS NULL"
        params = [self.service, label]
        if excluding_digest:
            sql += " AND token_sha256 != ?"
            params.append(excluding_digest)
        with connect(self.db_path) as conn:
            return conn.execute(sql, params).fetchone() is not None

File: node/src/store.ts
/**
 * The token store: schema, migrations and row access, on node:sqlite.
 *
 * Same schema as the Python package (`tokens` + `_schema_version`), so one store
 * file is readable by both runtimes. `ensureSchema` is the ONE creation /
 * migration routine — the CLI and the Express mount both call it, and two
 * independent creation paths would drift.
 *
 * No dependency: node:sqlite ships with Node (>=22.13 without a flag).
 */
import { chmodSync, existsSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { DatabaseSync } from "node:sqlite";

import { AmbiguousHandle, NoSuchHandle, TokenError } from "./errors.js";
import { handleOf, mintToken, tokenHash, validateHandle, validateRole, validateService } from "./hashing.js";

export const SCHEMA_VERSION = 2;

export const MAX_LIMIT = 500;
export const TOUCH_MAP_MAX = 5000;

const DDL = `
CREATE TABLE IF NOT EXISTS tokens (
  token_sha256 TEXT PRIMARY KEY,
  service      TEXT NOT NULL,
  label        TEXT NOT NULL,
  role         TEXT NOT NULL,
  priority     INTEGER,
  rate_limit   TEXT,
  expires_at   TEXT,
  created_at   TEXT NOT NULL,
  created_by   TEXT NOT NULL,
  last_used_at TEXT,
  revoked_at   TEXT,
  revoked_by   TEXT
);
CREATE INDEX IF NOT EXISTS idx_tokens_live ON tokens(service, revoked_at);
CREATE INDEX IF NOT EXISTS idx_tokens_handle ON tokens(service, substr(token_sha256, 1, 16));
CREATE UNIQUE INDEX IF NOT EXISTS idx_tokens_live_label
  ON tokens(service, label) WHERE revoked_at IS NULL;
CREATE TABLE IF NOT EXISTS _schema_version (
  id      INTEGER PRIMARY KEY CHECK (id = 1),
  version INTEGER NOT NULL
);
`;

// Columns added after v1, with the migration that introduces each.
// CREATE TABLE IF NOT EXISTS never alters an existing table, so a bare edit to
// DDL silently no-ops on adopted stores — every such column needs a step here.
const COLUMN_MIGRATIONS: Record<string, string> = {
  expires_at: "ALTER TABLE tokens ADD COLUMN expires_at TEXT",
};

/** One row of `tokens`. */
export interface TokenRow {
  token_sha256: string;
  service: string;
  label: string;
  role: string;
  priority: number | null;
  rate_limit: string | null;
  expires_at: string | null;
  created_at: string;
  created_by: string;
  last_used_at: string | null;
  revoked_at: string | null;
  revoked_by: string | null;
}

/**
 * Canonical timestamp: UTC, second precision, `+00:00` offset — byte-identical
 * to Python's `datetime.isoformat(timespec="seconds")`, because expiry is
 * compared as a string and a `Z`-form here vs a `+00:00`-form there would make
 * one runtime's tokens look expired to the other.
 */
export function utcNow(): string {
  return `${new Date().toISOString().slice(0, 19)}+00:00`;
}

export function normalizeExpiresAt(value: unknown): string | null {
  if (value == null) return null;
  if (typeof value !== "string" || value.trim() === "") {
    throw new Error("expires_at must be an ISO-8601 timestamp or null");
  }
  const m = /^(\d{4}-\d{2}-\d{2})[T ](\d{2}:\d{2}:\d{2})(?:\.\d+)?(Z|[+-]\d{2}:\d{2})?$/.exec(
    value.trim(),
  );
  if (!m) throw new Error(`expires_at is not an ISO-8601 timestamp: ${JSON.stringify(value)}`);
  const at = new Date(`${m[1]}T${m[2]}${m[3] ?? "+00:00"}`);
  if (Number.isNaN(at.getTime())) {
    throw new Error(`expires_at is not a valid timestamp: ${JSON.stringify(value)}`);
  }
  return `${at.toISOString().slice(0, 19)}+00:00`;
}

/** True iff `expiresAt` is set and not in the future. String compare: every
 * stored timestamp is normalised to the same UTC format. */
export function isExpired(expiresAt: string | null, now = utcNow()): boolean {
  if (!expiresAt) return false;
  return expiresAt <= now;
}

/** Best-effort chmod 0600 on the store file. The store holds digests, labels
 * and actors — not secrets — so this is hygiene, not a boundary; a failure must
 * not stop an app from starting. */
function restrictPermissions(path: string): void {
  for (const p of [path, `${path}-wal`, `${path}-shm`]) {
    try {
      if (existsSync(p)) chmodSync(p, 0o600);
    } catch {
      /* best effort */
    }
  }
}

function withDb<T>(path: string, fn: (db: DatabaseSync) => T): T {
  const db = new DatabaseSync(path);
  try {
    db.exec("PRAGMA journal_mode = WAL;");
    db.exec("PRAGMA busy_timeout = 5000;");
    db.exec("PRAGMA foreign_keys = ON;");
    return fn(db);
  } finally {
    db.close();
  }
}

function tableExists(db: DatabaseSync, name: string): boolean {
  const row = db
    .prepare("SELECT 1 FROM sqlite_master WHERE type='table' AND name = ?")
    .get(name);
  return row !== undefined;
}

function columnExists(db: DatabaseSync, table: string, column: string): boolean {
  const rows = db.prepare(`PRAGMA table_info(${table})`).all() as Array<{ name: string }>;
  return rows.some((r) => r.name === column);
}

export function ensureSchema(dbPath: string, { create }: { create: boolean }): void {
  if (!existsSync(dbPath)) {
    if (!create) {
      throw new TokenError(
        `${dbPath} does not exist. Pass --init to create it, and check the path ` +
          `matches the one the app passes to installTokens().`,
      );
    }
    mkdirSync(dirname(dbPath), { recursive: true });
  }
  withDb(dbPath, (db) => {
    const existing = tableExists(db, "_schema_version")
      ? (db.prepare("SELECT version FROM _schema_version WHERE id = 1").get() as
          | { version: number }
          | undefined)
      : undefined;
    if (existing && existing.version > SCHEMA_VERSION) {
      throw new TokenError(
        `store at ${dbPath} is schema v${existing.version}, newer than this package ` +
          `(v${SCHEMA_VERSION}). Refusing to downgrade — no DDL applied.`,
      );
    }
    if (existing && existing.version < SCHEMA_VERSION) {
      db.exec("BEGIN");
      try {
        if (existing.version < 2) {
          for (const [column, ddl] of Object.entries(COLUMN_MIGRATIONS)) {
            if (!columnExists(db, "tokens", column)) db.exec(ddl);
          }
        }
        db.exec("COMMIT");
      } catch (err) {
        db.exec("ROLLBACK");
        throw err;
      }
    }
    db.exec(DDL);
    // Upsert, not INSERT OR IGNORE: an upgraded store already has the id=1 row,
    // and ignoring it would leave the version stale forever.
    db.prepare(
      "INSERT INTO _schema_version (id, version) VALUES (1, ?)" +
        " ON CONFLICT(id) DO UPDATE SET version = excluded.version",
    ).run(SCHEMA_VERSION);
  });
  restrictPermissions(dbPath);
}

/** Service-scoped row access. Every statement carries `service`. */
export class Store {
  readonly dbPath: string;
  readonly service: string;
  readonly maxLimit: number;
  readonly touchMapMax: number;
  /** Monotonic ms of the last throttled touch / auth_failed emit, per key. */
  readonly lastTouch = new Map<string, number>();
  readonly lastAuthFailed = new Map<string, number>();

  constructor(dbPath: string, service: string, opts: { maxLimit?: number; touchMapMax?: number } = {}) {
    validateService(service);
    this.dbPath = dbPath;
    this.service = service;
    this.maxLimit = opts.maxLimit ?? MAX_LIMIT;
    this.touchMapMax = opts.touchMapMax ?? TOUCH_MAP_MAX;
  }

  /** Record `key` at `now`, evicting oldest-first past the bound. Delete-then-set
   * so a re-emitted key moves to the newest end instead of being evicted as the
   * oldest. */
  touchNote(map: Map<string, number>, key: string, now: number): void {
    map.delete(key);
    map.set(key, now);
    while (map.size > this.touchMapMax) {
      const oldest = map.keys().next().value;
      if (oldest === undefined) break;
      map.delete(oldest);
    }
  }

  mint(
    role: string,
    label: string,
    createdBy: string,
    extra: { priority?: number | null; rateLimit?: string | null; expiresAt?: string | null } = {},
  ): { raw: string; handle: string } {
    validateRole(role);
    const raw = mintToken(this.service, role);
    const handle = this.insertDigest(tokenHash(raw), role, label, createdBy, extra);
    return { raw, handle };
  }

  insertDigest(
    digest: string,
    role: string,
    label: string,
    createdBy: string,
    extra: { priority?: number | null; rateLimit?: string | null; expiresAt?: string | null } = {},
  ): string {
    validateRole(role);
    withDb(this.dbPath, (db) => {
      db.prepare(
        "INSERT INTO tokens (token_sha256, service, label, role, priority, rate_limit," +
          " expires_at, created_at, created_by) VALUES (?,?,?,?,?,?,?,?,?)",
      ).run(
        digest,
        this.service,
        label,
        role,
        extra.priority ?? null,
        extra.rateLimit ?? null,
        normalizeExpiresAt(extra.expiresAt ?? null),
        utcNow(),
        createdBy,
      );
    });
    return handleOf(digest);
  }

  lookupLive(digest: string, now = utcNow()): TokenRow | undefined {
    return withDb(this.dbPath, (db) =>
      db
        .prepare(
          "SELECT * FROM tokens WHERE token_sha256 = ? AND service = ?" +
            " AND revoked_at IS NULL AND (expires_at IS NULL OR expires_at > ?)",
        )
        .get(digest, this.service, now),
    ) as TokenRow | undefined;
  }

  findAny(digest: string): TokenRow | undefined {
    return withDb(this.dbPath, (db) =>
      db
        .prepare("SELECT * FROM tokens WHERE token_sha256 = ? AND service = ?")
        .get(digest, this.service),
    ) as TokenRow | undefined;
  }

  listRows(opts: { includeRevoked?: boolean; limit?: number; label?: string | null } = {}): {
    rows: TokenRow[];
    hasMore: boolean;
  } {
    const requested = opts.limit ?? this.maxLimit;
    const limit = Math.max(1, Math.min(Math.trunc(requested), this.maxLimit));
    let clause = opts.includeRevoked ? "" : " AND revoked_at IS NULL";
    const params: Array<string | number> = [this.service];
    if (opts.label != null) {
      clause += " AND label = ?";
      params.push(opts.label);
    }
    const rows = withDb(this.dbPath, (db) =>
      db
        .prepare(
          `SELECT * FROM tokens WHERE service = ?${clause} ORDER BY created_at, rowid LIMIT ?`,
        )
        .all(...params, limit + 1),
    ) as unknown as TokenRow[];
    return { rows: rows.slice(0, limit), hasMore: rows.length > limit };
  }

  resolveHandle(handle: string): TokenRow {
    validateHandle(handle);
    const rows = withDb(this.dbPath, (db) =>
      db
        .prepare("SELECT * FROM tokens WHERE service = ? AND substr(token_sha256,1,16) = ?")
        .all(this.service, handle),
    ) as unknown as TokenRow[];
    if (rows.length === 0) {
      throw new NoSuchHandle(`no token with handle ${JSON.stringify(handle)} in service ${JSON.stringify(this.service)}`);
    }
    if (rows.length > 1) {
      throw new AmbiguousHandle(
        `handle ${JSON.stringify(handle)} matches ${rows.length} tokens in service ` +
          `${JSON.stringify(this.service)}; use a longer prefix from the listing`,
      );
    }
    return rows[0];
  }

  /** Returns the row and whether it changed. The already-revoked guard is in the
   * UPDATE's WHERE clause, not a preceding read — a check-then-act pair lets two
   * concurrent revokes both pass and the second overwrite the first's revoked_by. */
  revoke(handle: string, revokedBy: string): { row: TokenRow; changed: boolean } {
    const row = this.resolveHandle(handle);
    const changed = withDb(this.dbPath, (db) => {
      const res = db
        .prepare(
          "UPDATE tokens SET revoked_at = ?, revoked_by = ?" +
            " WHERE token_sha256 = ? AND service = ? AND revoked_at IS NULL",
        )
        .run(utcNow(), revokedBy, row.token_sha256, this.service);
      return Number(res.changes) === 1;
    });
    return { row: this.resolveHandle(handle), changed };
  }

  updateRow(
    handle: string,
    patch: {
      label?: string | null;
      priority?: number | null | undefined;
      rateLimit?: string | null | undefined;
      expiresAt?: string | null | undefined;
    },
  ): TokenRow {
    const row = this.resolveHandle(handle);
    const sets: string[] = [];
    const params: Array<string | number | null> = [];
    if (patch.label != null) {
      sets.push("label = ?");
      params.push(patch.label);
    }
    if (patch.priority !== undefined) {
      sets.push("priority = ?");
      params.push(patch.priority);
    }
    if (patch.rateLimit !== undefined) {
      sets.push("rate_limit = ?");
      params.push(patch.rateLimit);
    }
    if (patch.expiresAt !== undefined) {
      sets.push("expires_at = ?");
      params.push(normalizeExpiresAt(patch.expiresAt));
    }
    if (sets.length > 0) {
      withDb(this.dbPath, (db) => {
        db.prepare(
          `UPDATE tokens SET ${sets.join(", ")} WHERE token_sha256 = ? AND service = ?`,
        ).run(...params, row.token_sha256, this.service);
      });
    }
    return this.resolveHandle(handle);
  }

  updatePolicy(digest: string, priority: number | null, rateLimit: string | null): void {
    withDb(this.dbPath, (db) => {
      db.prepare(
        "UPDATE tokens SET priority = ?, rate_limit = ? WHERE token_sha256 = ? AND service = ?",
      ).run(priority, rateLimit, digest, this.service);
    });
  }

  touchLastUsed(digest: string): void {
    withDb(this.dbPath, (db) => {
      db.prepare(
        "UPDATE tokens SET last_used_at = ? WHERE token_sha256 = ? AND service = ?",
      ).run(utcNow(), digest, this.service);
    });
  }

  liveLabelExists(label: string, opts: { excludingDigest?: string } = {}): boolean {
    let sql = "SELECT 1 FROM tokens WHERE service = ? AND label = ? AND revoked_at IS NULL";
    const params: string[] = [this.service, label];
    if (opts.excludingDigest) {
      sql += " AND token_sha256 != ?";
      params.push(opts.excludingDigest);
    }
    return (
      withDb(this.dbPath, (db) => db.prepare(sql).get(...params)) !== undefined
    );
  }
}
